pkg/kapp/diff/change_set_with_versioned_rs.go

```diff
@@ -238,7 +248,14 @@ func (ChangeSetWithVersionedRs) numOfResourcesToKeep(res ctlres.Resource) (int,
 		if numToKeep < 1 {
 			return 0, fmt.Errorf("Expected annotation '%s' value to be a >= 1", versionedResNumVersAnnKey)
 		}
+	} else if versionedRes.IsVersioned() && !versionedRes.IsTracked() {
+		if change.Op() == ChangeOpKeep {
+			numToKeep = 1
+		} else {
+			numToKeep = 0
+		}
```

Comment:
it's tricky to explain why this is necessary, i might need to think about adding a comment to the code why this necessary.

but for now I'll add it here:

Generally for name-hash-suffix CMs it's useless to keep old CMs, as we cannot properly order them there is not "last-X-versions" semantic. Thus, we always want to delete (all) existing/old ones. But we're facing a problem here when the CM does not change -- then there is no change that creates the CM but this method would still lead to the not-actually-old existing CM to get deleted.

Using the change to determine if the existing CM is old or, well, existing, I added to pass down the change that "produces" this CM -- if it's a "keep" change, it's an unchanged CM we want to keep, otherwise an old one we want to deleted.

Reply:
> it's tricky to explain why this is necessary

I actually was thinking of something similar when the we were discussing about this the last time, but I do think that there might be a better way of doing this... I will give it some more thought and let you know if I come up with something.